engine/runtime-parser/src/main/scala/org/enso/compiler/core/ir/LazyDiagnosticStorage.scala

```diff
@@ -0,0 +1,32 @@
+package org.enso.compiler.core.ir
+
+import org.enso.compiler.core.IR
+
+trait LazyDiagnosticStorage { self: IR =>
+
+  private[this] var _diagnostics: DiagnosticStorage = _
+
+  protected def diagnostics_=(diagnostics: DiagnosticStorage): Unit = {
+    assert(_diagnostics eq null)
+    _diagnostics = diagnostics
+  }
+
+  override def diagnostics: DiagnosticStorage = {
+    _diagnostics
+  }
+
+  override def diagnosticsList: List[Diagnostic] = {
```

Comment:
Oh, now I get it. You essentially have all those methods to deal with the possibility of `null` and not repeat the logic all over the place.

Reply:
I believe so. Thus in order to simplify **SPI** (e.g. implementations re-written by Dmitry in this PR), we have complicated [Client API](http://wiki.apidesign.org/wiki/ClientAPI). I see it as a completely opposite approach - the usage or `IR` should be simple. The implementation needn't be - API authors have to live thru their pain.